public/app/features/query/components/QueryEditorRow.tsx

```diff
@@ -450,9 +455,17 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
     const isHidden = !!query.hide;
 
     const hasEditorHelp = datasource?.components?.QueryEditorHelp;
+    const hasQueryLibrary = this.props.app === CoreApp.Explore && (config.featureToggles.queryLibrary || false);
 
     return (
       <>
+        {hasQueryLibrary && this.props.onBeginQuerySave !== undefined && (
+          <QueryOperationAction
```

Comment:
Is there be some indication that query is already in library?

Reply:
Not at the moment. We have the ability to save to query library from query history in Explore, but all it does is immediately hide the "save to query library" button after the save goes through in that component's state without persistence. If that's good enough we can absolutely do that here, but I don't know a good, scalable way to know one query is the same as one that already exists in the library. The query text isn't guaranteed to save properties in the same order, and searching the library off the query text would get really slow. There is a key for a query run but running the same query multiple times results in a different key every time, I believe. 

Is there something I'm missing?

Reply:
confused, so the saved query is not named or referenced by some name/id? or is that coming later?

Reply:
We have IDs for the queries that are saved in the query library. So if a query library entry exists for `SELECT * FROM table` , but then a couple weeks later someone goes into the same datasource and writes `SELECT * FROM table`, there is no great way to know that query is equivalent to one that is already saved. We can have something in app state where the button for that query row disappears immediately after saving, but the button would come back on refresh.